src/NUnitEngine/nunit.engine.tests/Services/AgentDatabaseTests.cs

```diff
@@ -0,0 +1,228 @@
+            Assert.That(process, Is.SameAs(DummyProcess));
+        }
+
+        [Test]
+        public static void AgentIsRunningWhenRegistered()
+        {
+            var database = new AgentDatabase();
+
+            database.Start(DummyAgent.Id, DummyProcess);
+            database.Register(DummyAgent);
+            Assert.That(database.IsAgentRunning(DummyAgent.Id, out var process), Is.True);
+            Assert.That(process, Is.SameAs(DummyProcess));
+        }
+
+        [Test]
+        public static void AgentIsNotRunningWhenTerminated()
+        {
+            var database = new AgentDatabase();
+
+            database.Start(DummyAgent.Id, DummyProcess);
+            database.Register(DummyAgent);
+            database.MarkTerminated(DummyAgent.Id);
+            Assert.That(database.IsAgentRunning(DummyAgent.Id, out _), Is.False);
+        }
+
+        [Test]
+        public static void ConcurrentOperationsDoNotCorruptState()
+        {
+            var database = new AgentDatabase();
+
+            RunActionConcurrently(() =>
+            {
+                for (var i = 0; i < 1000; i++)
+                {
+                    var id = Guid.NewGuid();
+
+                    Assert.That(database.IsAgentRunning(id, out _), Is.False);
+                    Assert.That(database.IsReady(id, out _), Is.False);
+
+                    database.Start(id, DummyProcess);
+                    Assert.That(database.IsAgentRunning(id, out _), Is.True);
+                    Assert.That(database.IsReady(id, out _), Is.False);
+
+                    database.Register(new DummyTestAgent(id));
+                    Assert.That(database.IsAgentRunning(id, out _), Is.True);
+                    Assert.That(database.IsReady(id, out _), Is.True);
+
+                    database.MarkTerminated(id);
+                    Assert.That(database.IsAgentRunning(id, out _), Is.False);
+                    Assert.That(database.IsReady(id, out _), Is.False);
+                }
+            }, threadCount: Environment.ProcessorCount);
+        }
+
+        private static void RunActionConcurrently(Action action, int threadCount)
+        {
+            var threads = new List<Thread>();
+            var exceptions = new List<Exception>();
+
+            for (var i = 0; i < threadCount; i++)
+            {
+                threads.Add(new Thread(() =>
+                {
+                    try
+                    {
+                        action.Invoke();
+                    }
+                    catch (Exception ex)
+                    {
+                        lock (exceptions)
+                            exceptions.Add(ex);
+                    }
+                }));
+            }
+
+            foreach (var thread in threads)
+                thread.Start();
+
+            foreach (var thread in threads)
+                thread.Join();
+
+            if (exceptions.Count != 0) throw exceptions[0];
```

Comment:
I was going to suggest `AggregateException` here, but apparently that's .NET 4 😢